README.md
# Open in Map 🗺️

[Github](https://github.com/Patrick64/vscode-open-in-map) | [Marketplace](https://marketplace.visualstudio.com/items?itemName=kirbydigital.openinmap) | [Linkedin](https://www.linkedin.com/in/patrick-woodcock-258371a6/)

**Open in Map** is a Visual Studio Code extension that allows developers to visualize geographical coordinates directly within the editor. By simply selecting latitude and longitude values in your code, you can open a map view in a split editor and see the exact locations marked with pins.

![Screenshot](./images/screenshot.png)

## Features 🚀

1. **Instant Map Visualization**: Select latitude and longitude values, and see them on a map instantly.
0. **Multiple pins** Adding additional lat/lngs will add to the existing map.
0. **Interactive Map**: Pan, zoom, and explore the map right within VS Code.
0. **Elegant Animations**: New markers are shown with an expanding circle.
0. **Robust matching**: As long as the lat/lngs are comma separated it will find within arrays and other formats eg `["-20.1235,-19.236", "-12.466,23.467"]` and `<loc>-20.1235, -19.236</loc><loc>-12.466, 23.467</loc>` will work.

## Installation 💻

1. Open the Extensions view in VS Code (`Ctrl+Shift+X`).
2. Search for "Open in Map."
3. Click "Install" to install the extension.
4. Reload VS Code to activate the extension.

## Usage 🌐

1. Highlight the latitude and longitude values in your code (e.g., `51.501476,-0.140634`).
2. Use the command palette (`Ctrl+Shift+P`) and search for "Open in map."
3. Enjoy the map view in a split editor with the selected coordinates marked.

## Acknowledgments 🙏

Special thanks to the Leaflet library for providing the mapping functionality. The maps are served from Open Street Map, please consider [donating](https://donate.openstreetmap.org/) to them to keep the servers running.

## Author 👨‍💻

Developed by Patrick Woodcock, a contract software developer in the UK. [Linkedin](https://www.linkedin.com/in/patrick-woodcock-258371a6/)

## Contributing 🤝

Contributions are welcome! Feel free to open an issue or submit a pull request.

## License 📄

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

CHANGELOG.md
# Change Log
All notable changes to the "open in map" extension will be documented in this file.

## Version 1.1

More robust pattern matching

## Version 1
- Initial release

src/test/suite/extension.test.ts
import * as assert from 'assert';
import * as vscode from 'vscode';
import * as sinon from 'sinon';
import { expect } from 'chai';
import {
  openMapInWebview,
  getMapHtml,
  convertStrToMarkers,
} from '../../openMapInWebview';
import * as openMapInWebviewModule from '../../openMapInWebview';

suite('Extension Test Suite', () => {
  vscode.window.showInformationMessage('Start all tests.');

  let createWebviewPanelStub: sinon.SinonStub;
  let onDidDisposeStub: sinon.SinonStub;
  let getMapHtmlStub: sinon.SinonStub;
  let asWebviewUriStub: sinon.SinonStub;
  let revealStub: sinon.SinonStub;
  let postMessageStub: sinon.SinonStub;

  setup(() => {
    // Stub the getMapHtml function to return a fixed string
    getMapHtmlStub = sinon
      .stub(openMapInWebviewModule, 'getMapHtml')
      .returns(Promise.resolve('<html></html>'));
    onDidDisposeStub = sinon.stub();
    postMessageStub = sinon.stub();
    revealStub = sinon.stub();
    asWebviewUriStub = sinon.stub().returns(vscode.Uri.parse('someUri'));
    createWebviewPanelStub = sinon
      .stub(vscode.window, 'createWebviewPanel')
      .returns({
        onDidDispose: onDidDisposeStub,
        reveal: revealStub,
        webview: {
          postMessage: postMessageStub,
          asWebviewUri: asWebviewUriStub,
        },
      } as unknown as vscode.WebviewPanel);
  });

  teardown(() => {
    sinon.restore();
  });

  test('openMapInWebview should throw an error if no valid latitude/longitude values are found', async () => {
    const selectedText = 'invalid text';
    const context = { subscriptions: [] } as unknown as vscode.ExtensionContext;

    try {
      await openMapInWebview(selectedText, context);
      assert.fail('Expected function to throw an error');
    } catch (err: any) {
      expect(err.message).to.exist;
    }
  });

  test('should create a webview panel with the correct arguments', async () => {
    const selectedText = '51.501476,-0.140634';
    const context = { subscriptions: [] } as unknown as vscode.ExtensionContext;

    await openMapInWebview(selectedText, context);

    expect(
      createWebviewPanelStub.calledWith(
        'openInMap',
        'Open in map',
        vscode.ViewColumn.Beside,
        {
          enableScripts: true,
          retainContextWhenHidden: true,
        }
      )
    ).to.be.true;

    assert(createWebviewPanelStub.calledOnce);
    assert(onDidDisposeStub.calledOnce);

    createWebviewPanelStub.restore();
  });

  test('Convert string to lat/lngs', () => {
    const input = ' -5.1, -1.1  51.1, 9.1 \n -51.2, -9.33  ';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([
      ['-5.1', '-1.1'],
      ['51.1', '9.1'],
      ['-51.2', '-9.33'],
    ]);
  });

  test('Convert empty string to empty array', () => {
    const input = '';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([]);
  });

  test('Convert string with invalid format to empty array', () => {
    const input = 'abc, def ghi, jkl';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([]);
  });

  test('Convert string with mixed valid and invalid format to array of valid pairs', () => {
    const input = '50.123,-3.543 abc, def \n 51.123, -2.543 ghi, jkl';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([
      ['50.123', '-3.543'],
      ['51.123', '-2.543'],
    ]);
  });

  // Test case: check for proper handling of input with leading/trailing spaces and different number of spaces.
  test('Convert string to lat/lngs with leading/trailing/multiple spaces', () => {
    const input =
      '   50.123, -3.543   51.123,   -2.543  \n\n   52.68,   -1.9876   ';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([
      ['50.123', '-3.543'],
      ['51.123', '-2.543'],
      ['52.68', '-1.9876'],
    ]);
  });

  // Test case: check for proper handling of input with only one pair.
  test('Convert string to lat/lngs with one pair', () => {
    const input = '50.123,-3.543';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([['50.123', '-3.543']]);
  });

  // Test case: check for proper handling of empty input.
  test('Convert string to lat/lngs with empty string', () => {
    const input = '';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([]);
  });

  // Test case: check for proper handling of input with only spaces and newlines.
  test('Convert string to lat/lngs with only spaces and newlines', () => {
    const input = ' \n \n ';
    const result = convertStrToMarkers(input);
    expect(result).to.eql([]);
  });
});


mapPanel.js
var mymap = L.map('mapid').setView([0, 0], 12);
L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
    attribution:
        '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
}).addTo(mymap);

let markers = [];

window.addEventListener('message', (event) => {
    const message = event.data; // The JSON data our extension sent
    switch (message.command) {
        case 'addMarkers':
            markers = [...markers, ...message.markers];
            message.markers.forEach((marker) => {
                addMarker(marker, mymap, L);
            });

            mymap.fitBounds(markers, { maxZoom: 12 });
            break;
    }
});

function addMarker(marker, mymap, L) {
    const newMarker = L.marker(marker).addTo(mymap);
    newMarker.bindTooltip(`${marker[0]},${marker[1]}`, {
        permanent: false,
    });

    // add a circle to show new markers
    const circleMarker = L.circleMarker(marker, {
        radius: 10,
        fillColor: '#ff7800',
        color: '#3287CC',
        weight: 2,
        opacity: 1,
        fillOpacity: 0,
    }).addTo(mymap);

    // Animate the marker
    let originalRadius = circleMarker.getRadius();
    let originalOpacity = circleMarker.options.fillOpacity;
    let expandInterval = setInterval(() => {
        let currentRadius = circleMarker.getRadius();
        if (currentRadius < originalRadius + 30) {
            circleMarker.setRadius(currentRadius + 1);
        } else {
            clearInterval(expandInterval);
            mymap.removeLayer(circleMarker);
        }
    }, 50);
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { openMapInWebview } from './openMapInWebview';

function openSelectionInMap(context: vscode.ExtensionContext) {
  // The code you place here will be executed every time your command is executed

  const editor = vscode.window.activeTextEditor;
  if (!editor) {
    vscode.window.showInformationMessage('Open in map: No editor is active.');
  } else {
    const selections = editor.selections;
    if (
      selections.length === 0 ||
      (selections.length === 1 && selections[0].isEmpty)
    ) {
      vscode.window.showInformationMessage('Open in map: Nothing selected.');
    } else {
      editor.edit(async (builder) => {
        try {
          const locations = [];
          for (const selection of selections) {
            if (!selection.isEmpty) {
              locations.push(editor.document.getText(selection));
            }
          }
          const locationsList = locations.join(' ');
          if (/^\s+$/.test(locationsList)) {
            vscode.window.showInformationMessage(
              'Open in map: Nothing selected.'
            );
          } else {
            await openMapInWebview(locations.join(' '), context);
          }
        } catch (err: any) {
          vscode.window.showInformationMessage('Open in map: ' + err.message);
        }
      });
    }
  }
}

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  // The command has been defined in the package.json file
  // Now provide the implementation of the command with  registerCommand
  // The commandId parameter must match the command field in package.json
  let disposable = vscode.commands.registerCommand(
    'openinmap.openSelectionInMap',
    () => {
      openSelectionInMap(context);
    }
  );

  context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() {}


src/openMapInWebview.ts
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';

let panel: vscode.WebviewPanel | null;

/**
 * Opens a map with the specified latitude and longitude markers in a split editor within VS Code.
 * 
 * @param selectedText - A string containing latitude and longitude values.
 * @param context - The extension context provided by VS Code.
 * @throws Will throw an error if the format of the selectedText is not correct.
 */
export async function openMapInWebview(selectedText: string, context: vscode.ExtensionContext): Promise<void> {
    const markers = convertStrToMarkers(selectedText);
    
    if (markers.length === 0) {
        throw new Error("No latitude/longitude values found. Format should be lat,lng.");
    }

    if (!panel) {
        initializePanel(context);
    }

    if (panel) {
        panel.webview.postMessage({ command: 'addMarkers', markers });
        panel.reveal();
    }
}

/**
 * Converts a string containing latitude and longitude values to an array of [lat, lng] pairs.
 * 
 * @param str - A string containing latitude and longitude values.
 * @returns An array of [lat, lng] pairs.
 */
export function convertStrToMarkers(str: string): string[][] {
    const regex = /(\-?\d+(\.\d+)?\s*,\s*\-?\d+(\.\d+)?)/g;
    const pairs = str.match(regex);

    return (pairs ?? []).map(latlng => latlng.split(',').map(num => num.trim()));
}

/**
 * Initializes the webview panel to display the map.
 * 
 * @param context - The extension context provided by VS Code.
 */
export async function initializePanel(context: vscode.ExtensionContext): Promise<void> {
    panel = vscode.window.createWebviewPanel(
        'openInMap', 
        'Open in map', 
        vscode.ViewColumn.Beside, 
        {
            enableScripts: true,
            retainContextWhenHidden: true
        }
    );

    panel.onDidDispose(() => {
        panel = null;
    }, null, context.subscriptions);

    panel.webview.html = await getMapHtml(panel);
}

/**
 * Generates the HTML content for the webview panel by replacing placeholders with actual script paths.
 * 
 * @param panel - The webview panel to generate the HTML content for.
 * @returns The HTML content as a string.
 */
export async function getMapHtml(panel: vscode.WebviewPanel): Promise<string> {
    const scriptPathOnDisk = vscode.Uri.file(path.join(__dirname, '../mapPanel.js'));
    const scriptUri = panel.webview.asWebviewUri(scriptPathOnDisk).toString();

    const mapHtmlPath = path.join(__dirname, '../mapPanel.html');
    let htmlContent = await fs.promises.readFile(mapHtmlPath, 'utf8');

    return htmlContent.replace(/%%scriptUri%%/g, scriptUri).replace(/%%cspSource%%/g, panel.webview.cspSource);
}


extension.js
const fs = require('fs').promises;
const path = require('path');

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');

let panel;

/**
 * This function is in vscode extension. It will open the given url in a split editor in vscode.
 *
 * @param   {[type]}  url  [url description]
 *
 * @return  {[type]}       [return description]
 */
async function openMapInASplitEditor(selectedText, context) {
    const markers = selectedText.split(/\s+/).filter(str => /^\-?\d+(\.\d+)?\,\-?\d+(\.\d+)?$/.test(str)).map(loc => loc.split(','));

    if (markers.length === 0) {
        throw new Error("No latitude/longitude values found. Format should be lat,lng eg `51.501476,-0.140634 51.2341098,-2.5815403`");
    }

    if (!panel) {
        panel = vscode.window.createWebviewPanel(
            'openInMap', // Identifies the type of the webview. Used internally
            'Open in map', // Title of the panel displayed to the user
            vscode.ViewColumn.Beside, // Editor column to show the new webview panel in.
            {
                // Enable scripts in the webview
                enableScripts: true,
                // Keep the webview's context even when it's not visible
                retainContextWhenHidden: true
            }
        );

        // Handle the onDidDispose event
        panel.onDidDispose(() => {
            panel = null;
        }, null, context.subscriptions);

        panel.webview.html = await getMapHtml();
    }




    panel.webview.postMessage({ command: 'addMarkers', markers });

    // Make the panel the active tab
    panel.reveal();
}

async function getMapHtml() {
    const mapHtmlPath = path.join(__dirname, 'mapPanel.html');
    return await fs.readFile(mapHtmlPath, 'utf8');
}


// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    // console.log('Congratulations, your extension "escape-regexp" is now active!');

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    let disposable = vscode.commands.registerCommand('extension.openSelectionInMap', function () {
        // The code you place here will be executed every time your command is executed

        // Display a message box to the user
        // vscode.window.showInformationMessage('Hello World!');

        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            vscode.window.showInformationMessage('Open in map: No editor is active.');
        } else {
            const selections = editor.selections;
            if (selections.length == 0 || (selections.length == 1 && selections[0].isEmpty)) {
                vscode.window.showInformationMessage('Open in map: Nothing selected.');
            } else {
                editor.edit(async (builder) => {
                    try {
                        const locations = [];
                        for (const selection of selections) {
                            if (!selection.isEmpty) {
                                locations.push(editor.document.getText(selection));
                            }
                        }
                        const locationsList = locations.join(' ');
                        if (/^\s+$/.test(locationsList)) {
                            vscode.window.showInformationMessage('Open in map: Nothing selected.');
                        } else {
                            await openMapInASplitEditor(locations.join(' '), context);
                        }
                    } catch (err) {
                        vscode.window.showInformationMessage('Open in map: ' + err.message);
                    }
                });

            }
        }
    });

    context.subscriptions.push(disposable);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;